### PR TITLE
Prevent privilege escalation

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

Cross-program invocations refer to the original transaction to enforce the `is_signer` and `is_writable` attributes of invoked instructions.  Doing so respects the original transaction's privileges but allows a privilege escalation between programs.  Even though the original transaction authorizes the behavior, it is a large foot gun because it relies on a careful review of the programs being invoked.

#### Summary of Changes

Enforce invoked instructions privileges against the calling instruction's privileges rather than the original transaction's.

Fixes #

### CODE REVIEW

File: programs/bpf/rust/invoke/src/lib.rs [hunk 1]
```diff
@@ -32,128 +36,166 @@ entrypoint!(process_instruction);
...
-        );
-    }
+            info!("Test nested invoke");
+            {
+                assert!(accounts[ARGUMENT_INDEX].is_signer);
+
+                **accounts[ARGUMENT_INDEX].lamports.borrow_mut() -= 5;
+                **accounts[INVOKED_ARGUMENT_INDEX].lamports.borrow_mut() += 5;
+
+                info!("Fist invoke");
```
Comment (garious) [inline programs/bpf/rust/invoke/src/lib.rs:134]: Fist?

File: programs/bpf_loader/src/syscalls.rs [hunk 2]
```diff
@@ -48,6 +49,8 @@ pub enum SyscallError {
     ProgramNotSupported,
     #[error("{0}")]
     InstructionError(InstructionError),
+    #[error("Piviledge escalation in cross-program invocation")]
+    PriviledgeEscalation,
```
Comment (garious) [inline programs/bpf_loader/src/syscalls.rs:53]: Seems like users would be confused by this error. Maybe "unauthorized something"?

Comment (CriesofCarrots) [reply programs/bpf_loader/src/syscalls.rs:53]: Also, Piviledge/Priviledge => Privilege

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs:53]: Heh, I'll rework it, sounds so cool though!

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs:53]: Thanks @CriesofCarrots 

Reworked...

Comment (CriesofCarrots) [reply programs/bpf_loader/src/syscalls.rs]: Just do a quick search for "priviledge". Looks like there are 14 other instances (const names, and one `info`)

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs]: Heh, thanks @CriesofCarrots, I think I'm spelling it as I say it ;-)

**Pull request merged**